### PR TITLE
feat: [PL-38941]: Handle account specific QPM and add metric to count number of external API requests

### PR DESCRIPTION
**Pull request by Raghav-Murali**
## Describe your changes

PR to support the below functionalities -
1. Handle account specific QPM limits.
2. Count the number of external API requests coming for an account. Currently there is no way to determine how many external API requests are coming and whether the requests are getting throttled appropriately.3. 

I have tested it in the local test env.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48628)

### CODE REVIEW

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java` line 68 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHokHSD_4Pqddn7nc2:-NXHokHSD_4Pqddn7nc3:b-kqtkfx) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java#L68)):*
> ```Java
>     ConfiguredLimit<StaticLimit> configuredLimit;
>     if (accountId != null) {
>       configuredLimit = limitConfigurationService.getOrDefault(accountId, ActionType.MAX_QPM_PER_MANAGER);
> ```

This logic does not look right if we don't have anything configured for specific account then we should return configured global limit for GLOBAL_ACCOUNT_ID

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java` line 67 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHogQgB4aRaKxg7dfM:-NXHogQgB4aRaKxg7dfN:b6pjfzz) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java#L67)):*
> ```Java
>     String accountId = apiKeyService.getAccountIdFromApiKey(key);
>     ConfiguredLimit<StaticLimit> configuredLimit;
>     if (accountId != null) {
> ```

This should be StringUtils.isNotEmpty

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java` line 331 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHo_T75yZU7P5NSx75:-NXHo_T75yZU7P5NSx76:bv6jop) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java#L331)):*
> ```Java
>   public String getAccountIdFromApiKey(String apiKey) {
>     try {
>       if (apiKey.contains(DELIMITER)) {
> ```

We should check for notEmpty

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/security/AuthenticationFilter.java` line 375 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHoHEoAw9B1MY65j6e:-NXHoHEoAw9B1MY65j6f:b1mi6t4) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/security/AuthenticationFilter.java#L375)):*
> ```Java
>     String key = containerRequestContext.getHeaderString(API_KEY_HEADER);
>     String accountId = apiKeyService.getAccountIdFromApiKey(key);
>     if (!accountId.isEmpty()) {
> ```

check for StringUtils.isNotEmpty instead and this check should be inside  externalApiMetricsService.recordApiRequestMetric

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/security/AuthenticationFilter.java` line 376 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHpXbo0wGblwpzhtYL:-NXHpXbo0wGblwpzhtYM:bmpvqja) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/security/AuthenticationFilter.java#L376)):*
> ```Java
>     String accountId = apiKeyService.getAccountIdFromApiKey(key);
>     if (!accountId.isEmpty()) {
>       externalApiMetricsService.recordApiRequestMetric(accountId, EXTERNAL_API_REQUEST_COUNT);
> ```

We should also publish requestPath

**Review by piyush-harness — COMMENTED** [review]
> Reviewed 4 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48628)* status: 8 file(s) to review. 0 of 1 approvals obtained. Waiting on team(s): del-team.

**Review by Raghav-Murali — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48628)* status: 9 file(s) to review. 6. unresolved discussions. 0 of 1 approvals obtained. Waiting on team(s): del-team. 

___
*[`400-rest/src/main/java/software/wings/security/AuthenticationFilter.java` line 375 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHoHEoAw9B1MY65j6e:-NXIBpVk6i6HYxzxnlD_:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/security/AuthenticationFilter.java#L375)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

check for StringUtils.isNotEmpty instead and this check should be inside  externalApiMetricsService.recordApiRequestMetric 
</blockquote></details>

Done
___
*[`400-rest/src/main/java/software/wings/security/AuthenticationFilter.java` line 376 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHpXbo0wGblwpzhtYL:-NXICkwa6Ys7ulwp2q-6:bb71kmx) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/security/AuthenticationFilter.java#L376)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

We should also publish requestPath 
</blockquote></details>

Okay
___
*[`400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java` line 331 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHo_T75yZU7P5NSx75:-NXICun0D5Nv1o-lHreT:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java#L331)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

We should check for notEmpty
</blockquote></details>

Done
___
*[`400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java` line 338 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHocF07l-kTk_WdNIW:-NXINOQxEAyOFssZ5916:b-2vwn3v) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java#L338)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Why catch RuntimeException?
</blockquote></details>

Have updated to handle just the IllegalArgumentException. This exception is being thrown if the API Key encoding is not correct.
___
*[`400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java` line 67 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHogQgB4aRaKxg7dfM:-NXIDdIyD5dt7rUavjiS:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java#L67)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

This should be StringUtils.isNotEmpty
</blockquote></details>

Done
___
*[`400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java` line 68 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHokHSD_4Pqddn7nc2:-NXIDezFDLq-q54DcMOL:bue35jy) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java#L68)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

This logic does not look right if we don't have anything configured for specific account then we should return configured global limit for GLOBAL_ACCOUNT_ID
</blockquote></details>

Oh yeah makes sense, will update it.

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger messagecheck

**Comment by Raghav-Murali** [comment]
trigger messagecheck

**Comment by Raghav-Murali** [comment]
trigger its

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java` line 338 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHocF07l-kTk_WdNIW:-NXMRiEG2uyhby9xoGkU:b-f29est) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java#L338)):*
<details><summary><i>Previously, Raghav-Murali wrote…</i></summary><blockquote>

Have updated to handle just the IllegalArgumentException. This exception is being thrown if the API Key encoding is not correct.
</blockquote></details>

ok but why do we need to handle and return null accountId, this will be a change in behavior then right?

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java` line 68 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHokHSD_4Pqddn7nc2:-NXMUoJ887mp28wa8CnJ:b-guumiy) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java#L68)):*
<details><summary><i>Previously, Raghav-Murali wrote…</i></summary><blockquote>

Oh yeah makes sense, will update it.
</blockquote></details>

can we not use the same method getOrDefault by passing accountId and then check if anything is not returned then invoke the same method with global account id

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java` line 68 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHokHSD_4Pqddn7nc2:-NXMV8wk1c6bWYL8D62e:b-h1pieg) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java#L68)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

can we not use the same method getOrDefault by passing accountId and then check if anything is not returned then invoke the same method with global account id
</blockquote></details>

Also add unit tests

**Review by Raghav-Murali — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48628)* status: 15 file(s) to review. 6. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java` line 338 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHocF07l-kTk_WdNIW:-NXMZoMF3Q0F0NKJeMl9:b-3wd9xc) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java#L338)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

ok but why do we need to handle and return null accountId, this will be a change in behavior then right?
</blockquote></details>

The behavior of this function is to return Null if the given API key is not of correct format. Like the existing Delimiter check which returns null accountId if there is no Delimiter. Similarly for this exception (API Key is not encoded properly) then we will return null accountId.  
___
*[`400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java` line 68 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHokHSD_4Pqddn7nc2:-NXMaeYZ9K_l9r5dgnt6:b-lgvck8) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ExternalApiRateLimitingServiceImpl.java#L68)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Also add unit tests
</blockquote></details>

I thought of doing it like that but the issue is getOrDefault will return a default limit `getDefaultLimit()` if it doesn't find anything for the given accountId. That is why I added a new method to try and get limit for the given accountId else use the existing one.

I will add unit test for it.

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java` line 338 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHocF07l-kTk_WdNIW:-NXMpjwh2KKTslS2Esfo:b6gb7xc) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java#L338)):*
<details><summary><i>Previously, Raghav-Murali wrote…</i></summary><blockquote>

The behavior of this function is to return Null if the given API key is not of correct format. Like the existing Delimiter check which returns null accountId if there is no Delimiter. Similarly for this exception (API Key is not encoded properly) then we will return null accountId.  
</blockquote></details>

Before it used to throw IllegalArgumentException in case of wrong encoding correct?

**Comment by piyush-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java` line 338 at r1](https://reviewable.io/reviews/harness/harness-core/48628#-NXHocF07l-kTk_WdNIW:-NXMq1Fp3B_ErWUOE4sX:bqejuyy) ([raw file](https://github.com/harness/harness-core/blob/a53fff5868e31a54e377ee4fbf84b543c327e82f/400-rest/src/main/java/software/wings/service/impl/ApiKeyServiceImpl.java#L338)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Before it used to throw IllegalArgumentException in case of wrong encoding correct?
</blockquote></details>

I am talking about ^^^ this behavior change, It was terminating the thread by throwing a RuntimeException and now it will return Null AccountId.

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Review by piyush-harness — APPROVED** [review]
> Reviewed 5 of 12 files at r1, 9 of 10 files at r3, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48628)* status: 16 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): del-team.

File: 400-rest/src/test/java/io/harness/limits/configuration/LimitConfigurationServiceMongoIntegrationTest.java [hunk 1]
```diff
@@ -120,4 +124,42 @@ public void testUpsertForRateLimits() {
     fetchedValue = configuredLimitService.get(accountId, CREATE_APPLICATION);
     assertThat(fetchedValue).isEqualTo(cl);
   }
+
+  @Test
+  @Owner(developers = RAGHAV_MURALI)
+  @Category(DeprecatedIntegrationTests.class)
+  @Ignore("skipping the integration test")
+  public void testgetOrDefaultToGlobal_ReturnsGivenAccountLimit() {
```
Comment (gn-harness) [inline 400-rest/src/test/java/io/harness/limits/configuration/LimitConfigurationServiceMongoIntegrationTest.java:132]: nit: account and global limit both set - add this to test name

Comment (gn-harness) [reply 400-rest/src/test/java/io/harness/limits/configuration/LimitConfigurationServiceMongoIntegrationTest.java:132]: and for the next test, just the global limit set.

something like: testgetOrDefaultToGlobal_onlyGlobalLimitSet_returnsGlobalAccountLimit

**Review by gn-harness — APPROVED** [review]

**Pull request merged**